### DISCUSSION TITLE
Spawn magick from a C program and stay resident

### DISCUSSION THREAD

**Discussion by danielwlewis** [ImageMagick/ImageMagick#6473] [Help]
### ImageMagick version

7.1.1-12

### Operating system, version and so on

Windows 10

### Description

I have a multi-threaded C program that processes a large number of JPG files and uses magick to create thumbnails. Currently, the program enqueues the following three operations for each photo in a list:

1. Request one thread to copy the photo from a remote fileserver to the local disk
2. Request a second thread to extract EXIF information from the photo
3. Request a third thread to  create an HTML page that embeds the photo

Once the threads have completed these operations for all photos, the C program then spawns magick, giving it a text file containing one command per line for each photo in the list. The program spawns magick using _P_NOWAIT so that the creation of thumbnails can be overlapped with other (additional and unrelated) computation.

However, rather than waiting for the threads to complete before giving magick the entire list of JPGs to process all at once, I'd rather initiate all <b>FOUR</b> operations related to each photo at the same time. Specifically I want to enqueue four operations to be performed for each photo, and repeat this sequence for each photo in the list:

1. Request one thread to copy the photo from a remote fileserver to the local disk
2. Request a second thread to extract EXIF information from the photo
3. Request a third thread to create an HTML page that embeds the photo
4. Request a fourth thread to create a thumbnail using magick

This requires that the C program invoke magick as a separate process, and then send that process a sequence of commands (one per thumbnail to be created). After all the photos have been processed, I could use magick's process handle to terminate magick.

Does anyone know if there is an option to keep magick resident and then send it commands?  (BTW, Phil Harvey's program [exiftool ](https://exiftool.org/) has a similar capability that I use to extract EXIF data.)

Thanks!
Dan

**Comment by snibgo** [comment]
Your arrangement of threads seems strange, as the second thread can't start until the first has finished, etc.

> Does anyone know if there is an option to keep magick resident and then send it commands?

There isn't. I have written a fairly simple C program called `imserv.c` that waits for a text line, then executes it as a `magick` command, and repeats. See [Client-server](http://im.snibgo.com/cliserv.htm)

**Comment by urban-warrior** [comment]
Try `magick-script -`.  It will wait for commands that you pipe to it, execute the command, and then wait for the next.  To exit, send the `-exit` option.  You can also wait on other file descriptors, e.g., fd:5,  if you prefer rather than stdin.  See https://imagemagick.org/script/magick-script.php.

**Reply by danielwlewis** [reply]
Perfect!  Thank you!

Dan

**Comment by danielwlewis** [comment]
snibgo,

Thanks for the response. I'll take a look at your code. It sounds like what I'm looking for.

BTW, I agree that my description of my software would lead you to believe that there's a sequencing problem, but I tried to use brevity rather than providing too much detail. In fact, the EXIF extraction both waits for the copying from the file server to complete. Embedding the image in the webpage is simply a file reference and can proceed without waiting.

Best,
Dan

**Comment by snibgo** [comment]
See also `magick-script`. I had forgotten about that. That is slightly different to my `imserv.c`. `magick-script` waits for one long command, terminated by `-exit`. The command can be split across multiple lines and each line of the long command will be executed as it is received. `imserv.c` waits for one command (in a single line, or using line-continuation characters) which it executes immediately, then waits for another command which it executes immediately, etc.

**Reply by danielwlewis** [reply]
snibgo,

'Sorry, but I'm not sure I understand the difference if both imserv and magick-script **"execute each command immediately"**. Or are you saying that nothing happens with magick-script until it receives the "-exit" line, but that imserv responds at each end of line character?

Dan

**Comment by snibgo** [comment]
Sorry, I didn't explain it well. No, that's not what I mean.

Its about the difference between one single command and multiple independent commands, and also about individual options within a command.

Ordinary CLI `magick` processes and executes commands one option at a time. It doesn't "look ahead" within the command. If some option in the command fails, so the command as a whole fails, any previous options will have been executed. For example (Windows command):
```
magick -read toes.png -write x.png -write Z:y.png NULL:
```
I don't have a drive Z:, so that write fails, but the first `-write` has already succeeded.

And that is how "`magick-script-`" works, option by option. Type stuff in and, as soon as it has read one or more options, it executes those options. Overall, we give it a single command. So we might have "`define option:VAR=123`", and `VAR` will be available later in the command. When we exit "`magick-script-`", the single command is finished.

`imserv.c` is different. It doesn't execute options until it has the complete command. The command can be split between lines, but no options are executed until the line is complete. But it can execute multiple independent commands, and setting `VAR` in one command has no visibility in another.

(I don't claim that either method is better than the other, just different.)

**Reply by danielwlewis** [reply]
Thanks for clarifying. I ultimately went with "magick -script". The code I already had was using the -script option, but I didn't know about the "-exit" command or that magick would remain resident as commands were written to stdin using a pipe.

It also simplified my code considerably because I needed to redirect stdout and stderr to a file, and it was much easier to do that with popen since it calls the command line interpreter. The _spawn function I had been using does not, so I had to do the redirection separately in C before calling _spawn and then restore stdout and stderr after.

Thanks again!
Dan
